### ISSUE TITLE
Movable ssl::stream

### ISSUE THREAD

**Issue by 2underscores-vic** [chriskohlhoff/asio#124]
Is there any reason why `ssl::stream` is still not movable? As far as I can see, the feature was requested 2 years ago here https://svn.boost.org/trac/boost/ticket/9724

**Comment by vinniefalco** [comment]
I'm about to write my own wrapper that makes this thing movable.

**Comment by vinniefalco** [comment]
I'm thinking about writing this wrapper:
```
class ssl_socket : public stream_base
{
  using socket_type = boost::asio::ip::tcp::socket;
  using stream_type = boost::asio::ssl::stream<socket_type>;
  std::unique_ptr<stream_type> st_;

public:
  using impl_struct = stream_type::impl_struct;
  // ...

  // This would have to be constrained
  template<class... Args>
  explicit
  ssl_socket(Args&&... args)
    : st_(new stream_type{std::forward<Args>(args)...})
  {
  }

  ssl_socket(ssl_socket&& other)
    : st_(other.st_)
  {
  }

  //...

  void set_verify_depth(int depth)
  {
    st_->set_verify_depth(depth);
  }

  // ...

  template<class MutableBufferSequence>
  std::size_t
  read_some(MutableBufferSequence const& buffers,
      boost::system::error_code& ec)
  {
    return st_->read_some(buffers, ec);
  }

  //...
};
```

What do you think about this?

**Comment by vinniefalco** [comment]
@2underscores-vic This compiles, but is untested:
//
// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//

#ifndef BEAST_SSL_SOCKET_HPP
#define BEAST_SSL_SOCKET_HPP

#include <boost/asio/ip/tcp.hpp>
#include <boost/asio/ssl/stream.hpp>

namespace beast {

class ssl_socket
    : public boost::asio::ssl::stream_base
{
    using socket_type = boost::asio::ip::tcp::socket;
    using stream_type = boost::asio::ssl::stream<socket_type>;

    std::unique_ptr<stream_type> p_;

public:
    /// The native handle type of the SSL stream.
    using native_handle_type = stream_type::native_handle_type;

    /// Structure for use with deprecated impl_type.
    using impl_struct = stream_type::impl_struct;

    /// (Deprecated: Use native_handle_type.) The underlying implementation type.
    using impl_type = stream_type::impl_type;

    /// The type of the next layer.
    using next_layer_type = stream_type::next_layer_type;

    /// The type of the lowest layer.
    using lowest_layer_type = stream_type::lowest_layer_type;

    ssl_socket(ssl_socket&&) = default;
    ssl_socket(ssl_socket const&) = delete;
    ssl_socket& operator=(ssl_socket&&) = default;
    ssl_socket& operator=(ssl_socket const&) = delete;

    ssl_socket(boost::asio::io_service& ios,
            boost::asio::ssl::context& ctx)
        : p_(new stream_type{ios, ctx})
    {
    }

    boost::asio::io_service&
    get_io_service()
    {
        return p_->get_io_service();
    }

    native_handle_type
    native_handle()
    {
        return p_->native_handle();
    }

    impl_type
    impl()
    {
        return p_->impl();
    }

    next_layer_type const&
    next_layer() const
    {
        return p_->next_layer();
    }

    next_layer_type&
    next_layer()
    {
        return p_->next_layer();
    }

    lowest_layer_type&
    lowest_layer()
    {
        return p_->lowest_layer();
    }

    lowest_layer_type const&
    lowest_layer() const
    {
        return p_->lowest_layer();
    }

    void
    set_verify_mode(boost::asio::ssl::verify_mode v)
    {
        p_->set_verify_mode(v);
    }

    boost::system::error_code
    set_verify_mode(boost::asio::ssl::verify_mode v,
        boost::system::error_code& ec)
    {
        return p_->set_verify_mode(v, ec);
    }

    void
    set_verify_depth(int depth)
    {
        p_->set_verify_depth(depth);
    }

    boost::system::error_code
    set_verify_depth(
        int depth, boost::system::error_code& ec)
    {
        return p_->set_verify_depth(depth, ec);
    }

    template<class VerifyCallback>
    void
    set_verify_callback(VerifyCallback callback)
    {
        p_->set_verify_callback(callback);
    }

    template<class VerifyCallback>
    boost::system::error_code
    set_verify_callback(VerifyCallback callback,
        boost::system::error_code& ec)
    {
        return p_->set_verify_callback(callback, ec);
    }

    void
    handshake(handshake_type type)
    {
        p_->handshake(type);
    }

    boost::system::error_code
    handshake(handshake_type type,
        boost::system::error_code& ec)
    {
        return p_->handshake(type, ec);
    }

    template<class ConstBufferSequence>
    void
    handshake(
        handshake_type type, ConstBufferSequence const& buffers)
    {
        p_->handshake(type, buffers);
    }

    template<class ConstBufferSequence>
    boost::system::error_code
    handshake(handshake_type type,
        ConstBufferSequence const& buffers,
            boost::system::error_code& ec)
    {
        return p_->handshake(type, buffers, ec);
    }

    template<class HandshakeHandler>
    BOOST_ASIO_INITFN_RESULT_TYPE(HandshakeHandler,
        void(boost::system::error_code))
    async_handshake(handshake_type type,
        BOOST_ASIO_MOVE_ARG(HandshakeHandler) handler)
    {
        return p_->async_handshake(type,
            BOOST_ASIO_MOVE_CAST(HandshakeHandler)(handler));
    }

    template<class ConstBufferSequence, class BufferedHandshakeHandler>
    BOOST_ASIO_INITFN_RESULT_TYPE(BufferedHandshakeHandler,
        void (boost::system::error_code, std::size_t))
    async_handshake(handshake_type type, ConstBufferSequence const& buffers,
        BOOST_ASIO_MOVE_ARG(BufferedHandshakeHandler) handler)
    {
        return p_->async_handshake(type, buffers,
            BOOST_ASIO__MOVE_CAST(BufferedHandshakeHandler)(handler));
    }

    void
    shutdown()
    {
        p_->shutdown();
    }

    boost::system::error_code
    shutdown(boost::system::error_code& ec)
    {
        return p_->shutdown(ec);
    }

    template<class ShutdownHandler>
    BOOST_ASIO_INITFN_RESULT_TYPE(ShutdownHandler,
        void (boost::system::error_code))
    async_shutdown(BOOST_ASIO_MOVE_ARG(ShutdownHandler) handler)
    {
        return p_->async_shutdown(
            BOOST_ASIO_MOVE_CAST(ShutdownHandler)(handler));
    }

    template<class ConstBufferSequence>
    std::size_t
    write_some(ConstBufferSequence const& buffers)
    {
        return p_->write_some(buffers);
    }

    template<class ConstBufferSequence>
    std::size_t
    write_some(ConstBufferSequence const& buffers,
        boost::system::error_code& ec)
    {
        return p_->write_some(buffers, ec);
    }

    template<class ConstBufferSequence, class WriteHandler>
    BOOST_ASIO_INITFN_RESULT_TYPE(WriteHandler,
        void (boost::system::error_code, std::size_t))
    async_write_some(ConstBufferSequence const& buffers,
        BOOST_ASIO_MOVE_ARG(WriteHandler) handler)
    {
        return p_->async_write_some(buffers,
            BOOST_ASIO_MOVE_CAST(WriteHandler)(handler));
    }

    template<class MutableBufferSequence>
    std::size_t
    read_some(MutableBufferSequence const& buffers)
    {
        return p_->read_some(buffers);
    }

    template<class MutableBufferSequence>
    std::size_t
    read_some(MutableBufferSequence const& buffers,
        boost::system::error_code& ec)
    {
        return p_->read_some(buffers, ec);
    }

    template<class MutableBufferSequence, class ReadHandler>
    BOOST_ASIO_INITFN_RESULT_TYPE(ReadHandler,
        void(boost::system::error_code, std::size_t))
    async_read_some(MutableBufferSequence& buffers,
        BOOST_ASIO_MOVE_ARG(ReadHandler) handler)
    {
        return p_->async_read_some(buffers,
            BOOST_ASIO_MOVE_CAST(ReadHandler)(handler))
    }
};

} // beast

#endif

**Comment by 2underscores-vic** [comment]
Yes, I would use the same workaround - allocation on heap, but in my project I'm trying to minimize heap allocations, so it used only on highest levels of abstraction.

BTW, move operations can be easily added to `ssl::stream_base` (as far as I cant see). And `ssl::stream` object consists of `ssl::stream_base` + socket. Is there any thechnical difficulties to do this?

**Comment by vinniefalco** [comment]
> in my project I'm trying to minimize heap allocations

Have you looked at the implementation of `ssl::stream`? There are allocations everywhere, through `std::vector`, and inside OpenSSL itself. A single, additional allocation for each connection is not going to hurt.

> Is there any thechnical difficulties to do this

I don't know. You might look at the declaration for `ssl::detail::engine` and `ssl::detail::stream_core`. I assume there's a good reason that `ssl::stream` is not movable, considering that other types such as `basic_socket` are movable.

You're proposing changing the source code to Boost.Asio. That might work for your application but its  not really a good solution to force everyone who wants to use some piece of code to modify their installation of Boost.

**Comment by 2underscores-vic** [comment]
> You're proposing changing the source code to Boost.Asio
No, I use standalone Asio 1.11. And why others have to modify their code if they don't/can't use moving today?

**Comment by 2underscores-vic** [comment]
> BTW, move operations can be easily added to `ssl::stream_base`

Sorry, I meant `detail::stream_core`. Data members are:

```
asio::steady_timer/asio::deadline_timer pending_read_;
asio::steady_timer/asio::deadline_timer pending_write_;
std::vector<unsigned char> output_buffer_space_;
const asio::mutable_buffers_1 output_buffer_;
std::vector<unsigned char> input_buffer_space_;
const asio::mutable_buffers_1 input_buffer_;
asio::const_buffer input_;
```
`const` members are not move assignable but move copyable/movable. Can we just make them non-const?
@chriskohlhoff answer something.

**Comment by vinniefalco** [comment]
> why others have to modify their code if they don't/can't use moving today?

I can't add a feature to Beast (my HTTP library) that requires users to modify their Asio sources.
https://github.com/vinniefalco/Beast/

**Comment by 2underscores-vic** [comment]
I don't understand how this feature request to Asio relates to Beast or any other library...

**Comment by trivigy** [comment]
This feature request has been open for almost two years? I am using beast library and the immovability of the ssl::stream makes it super hard to pass a stream that has a handshake initiated on it into other classes. For example: handshare a stream in http class and hand it off to websocket class.

Any update on the progress would be much appreciated.

**Comment by trivigy** [comment]
Hey @vinniefalco thanks for sharing. I went through that code example with a fine comb. Really cool example. I ended up refactoring the code and using the ssl_stream but instead of all the generics and separate classes I used only two classes (Http, Websocket) and ended used boost::variant to pass both plain and secured sockets around.

**Comment by omartijn** [comment]
It's nice that we have a working example emulating a proper, movable ssl::stream. It is still confusing for developers though - you have to really search for how to do this instead of it "just" working.

Reading the comments on this issue, the only reason it seems we don't have movable streams is that the buffers are made const. I think the added value we get from those members being const in no way outweighs the benefits of having movable streams.

I would really like to hear @chriskohlhoff take on this. If he offers no objections, I'd consider making a PR to fix this once and for all.

**Comment by vinniefalco** [comment]
Beast offers an `ssl_stream` which is not only movable but overcomes a performance limitation of asio's `ssl::stream` when performing writes with buffer sequences having length greater than one: https://github.com/boostorg/asio/issues/100

Beast's `ssl_stream`: https://github.com/boostorg/beast/blob/develop/include/boost/beast/_experimental/core/ssl_stream.hpp#L55

This will become an "official" interface in an upcoming Boost release, but it can be relied upon today. Also note that Boost.Beast 1.70 will ship `beast::tcp_stream` which is a stream that has built-in timeouts. You can declare

```
beast::ssl_stream<beast::tcp_stream<net::io_context::strand>> stranded_stream(ioc);

// or, without the strand:
beast::ssl_stream<beast::tcp_stream<net::io_context::executor_type>> stream(ioc);
```

To get a stream with these features:
* Uses `tcp::socket` at the bottom
* Supports configurable built-in timeouts:
  - `beast::error::timeout` delivered to your completion handler
* Supports configurable bandwidth management (limits on reading and writing)
* Built-in `strand` (optional), never have to call `bind_executor` again
* Workaround for https://github.com/boostorg/asio/issues/100

**Comment by maddanio** [comment]
did you propose this to asio? if not, why not. I mean: great for the workaround!! but seems a bit messy to have to do this, especially since they are both boost libraries.

**Comment by vinniefalco** [comment]
> if not, why not

What would be the difference between being in beast or being in asio?

**Comment by omartijn** [comment]
> What would be the difference between being in beast or being in asio?

Visibility. When you think about "how do I sent something over this TCP socket with a timeout", the most logical place to look for something like this is asio. You won't find it there and might conclude you have to write it all on your own.

The things you write here are very useful. It'd be a shame if people don't use them because they don't find them.

**Comment by maddanio** [comment]
or in the original case for which this ticket was opened if the ssl stream in asio where to be made movable then users of asio, and also would directly pick up the movable streams and have less head aches. also they would not have to depend on beast to use it.

**Comment by vinniefalco** [comment]
beast's `ssl_stream` contains a workaround for the lack of scatter/gather I/O in `ssl::stream`. I don't think that workaround is appropriate in ASIO.

**Comment by maddanio** [comment]
How so?

**Comment by 2underscores-vic** [comment]
Resolved in https://github.com/chriskohlhoff/asio/commit/658614dca67896c0cec896efcc089f5028de1170?

**Comment by omartijn** [comment]
Oh, this looks very promising indeed! @vinniefalco Would you take a PR that switches out the `std::unique_ptr<stream_type` for a simple `stream_type`?

I guess to be properly compatible with standalone asio (which might have an older version) we could simply make a type trait that checks whether or not `stream_type` is move-constructible and assignable.

**Comment by vinniefalco** [comment]
The latest Asio makes `beast::ssl_stream` obsolete, so no.

**Comment by 2underscores-vic** [comment]
> Resolved in [658614d](https://github.com/chriskohlhoff/asio/commit/658614dca67896c0cec896efcc089f5028de1170)?

No.
1) Move constructor isn't `noexcept`.
2) No move-assignment at all.
